### ISSUE TITLE
Lost parts that aren't really lost => replication queue grows very fast

### ISSUE THREAD

**Issue by cyberhuman** [ClickHouse/ClickHouse#21361]
**Describe the bug**
We have noticed the number of nodes in ZK is growing very fast:
![image](https://user-images.githubusercontent.com/979624/109590234-93a62b00-7b46-11eb-9d36-3d2fc7701ee3.png)

After checking, it appears they correspond to the replication queue in CH:
```
SELECT
    count(*),
    type
FROM system.replication_queue
GROUP BY type

Query id: 22b47a8a-a637-4e02-80a2-e5f4afe512da

┌─count()─┬─type──────────┐
│  146833 │ REPLACE_RANGE │
│      47 │ MERGE_PARTS   │
└─────────┴───────────────┘
```

When checking further, these are errors because of missing parts, starting several days ago:
```
SELECT * EXCEPT database EXCEPT table
FROM system.replication_queue
WHERE type = 'REPLACE_RANGE'
ORDER BY create_time ASC
LIMIT 5

Query id: 94cd3b60-2a0d-4e10-9e15-ae781189a9dd

┌─replica_name─┬─position─┬─node_name────────┬─type──────────┬─────────create_time─┬─required_quorum─┬─source_replica─┬─new_part_name─┬─parts_to_merge─┬─is_detach─┬─is_currently_executing─┬─num_tries─┬─last_exception───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───last_attempt_time─┬─num_postponed─┬─postpone_reason─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬──last_postpone_time─┬─merge_type─┐
│ 1            │   108617 │ queue-0000107972 │ REPLACE_RANGE │ 2021-02-27 00:11:54 │               0 │ 0              │               │ []             │         0 │                      0 │         2 │ Code: 234, e.displayText() = DB::Exception: Not found part 15-4-0_3021_3021_0 (or part covering it) neither source table neither remote replicas (version 21.2.3.15) │ 2021-03-02 00:00:57 │        788022 │ Not executing log entry queue-0000107972 of type REPLACE_RANGE for part  because another DROP_RANGE or REPLACE_RANGE entry are currently executing. │ 2021-03-02 03:09:13 │            │
│ 1            │   108619 │ queue-0000107977 │ REPLACE_RANGE │ 2021-02-27 00:11:55 │               0 │ 0              │               │ []             │         0 │                      0 │         2 │ Code: 234, e.displayText() = DB::Exception: Not found part 5-4-0_3007_3007_0 (or part covering it) neither source table neither remote replicas (version 21.2.3.15)  │ 2021-03-02 00:00:58 │        788017 │ Not executing log entry queue-0000107977 of type REPLACE_RANGE for part  because another DROP_RANGE or REPLACE_RANGE entry are currently executing. │ 2021-03-02 03:09:13 │            │
│ 1            │   108620 │ queue-0000107979 │ REPLACE_RANGE │ 2021-02-27 00:11:55 │               0 │ 0              │               │ []             │         0 │                      0 │         2 │ Code: 234, e.displayText() = DB::Exception: Not found part 79-4-0_2915_2915_0 (or part covering it) neither source table neither remote replicas (version 21.2.3.15) │ 2021-03-02 00:00:58 │        788017 │ Not executing log entry queue-0000107979 of type REPLACE_RANGE for part  because another DROP_RANGE or REPLACE_RANGE entry are currently executing. │ 2021-03-02 03:09:13 │            │
│ 1            │   108618 │ queue-0000107974 │ REPLACE_RANGE │ 2021-02-27 00:11:55 │               0 │ 0              │               │ []             │         0 │                      0 │         2 │ Code: 234, e.displayText() = DB::Exception: Not found part 23-4-0_3027_3027_0 (or part covering it) neither source table neither remote replicas (version 21.2.3.15) │ 2021-03-02 00:00:58 │        788017 │ Not executing log entry queue-0000107974 of type REPLACE_RANGE for part  because another DROP_RANGE or REPLACE_RANGE entry are currently executing. │ 2021-03-02 03:09:13 │            │
│ 1            │   108622 │ queue-0000107981 │ REPLACE_RANGE │ 2021-02-27 00:11:55 │               0 │ 0              │               │ []             │         0 │                      0 │         2 │ Code: 234, e.displayText() = DB::Exception: Not found part 41-4-0_3019_3019_0 (or part covering it) neither source table neither remote replicas (version 21.2.3.15) │ 2021-03-02 00:00:59 │        788017 │ Not executing log entry queue-0000107981 of type REPLACE_RANGE for part  because another DROP_RANGE or REPLACE_RANGE entry are currently executing. │ 2021-03-02 03:09:13 │            │
└──────────────┴──────────┴──────────────────┴───────────────┴─────────────────────┴─────────────────┴────────────────┴───────────────┴────────────────┴───────────┴────────────────────────┴───────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────┴───────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────┴────────────┘

5 rows in set. Elapsed: 2.441 sec. Processed 146.88 thousand rows, 74.27 MB (60.17 thousand rows/s., 30.42 MB/s.) 
```

When checking the parts in `system.part_log`, they don't exist on the host, but exist on the replica.

Local host: no mention of part 15-4-0_3021_3021_0 in `system.part_log`, but there is a repeating error in the CH log:
```
254526:2021.02.27 00:12:09.282277 [ 118199 ] {} <Information> database.table_local (3bc87ec4-fcc2-4bd2-810a-3ffc12760885): DB::Exception: Not found part 15-4-0_3021_3021_0 (or part covering it) neither source table neither remote replicas
```

Remote host: no CH logs related to part 15-4-0_3021_3021_0, but there is a row in the `system.part_log` and the part is still there in `system.parts`:
```
SELECT * EXCEPT query_id EXCEPT database EXCEPT table
FROM system.part_log
WHERE (database = 'database') AND (part_name = '15-4-0_3021_3021_0')
ORDER BY event_time ASC

Query id: 2ddc47f0-7b48-443a-bbd1-c288f0ebde71

┌─event_type─┬─event_date─┬──────────event_time─┬─duration_ms─┬─part_name──────────┬─partition_id─┬─path_on_disk──────────────────────────────────────────────────────────────────────────┬─rows─┬─size_in_bytes─┬─merged_from─┬─bytes_uncompressed─┬─read_rows─┬─read_bytes─┬─peak_memory_usage─┬─error─┬─exception─┐
│ NewPart    │ 2021-02-27 │ 2021-02-27 00:11:54 │         569 │ 15-4-0_3021_3021_0 │ 15-4-0       │ /disk12/clickhouse/store/832/83200d96-3224-40dc-83de-ab7adedbf1ec/15-4-0_3021_3021_0/ │   77 │       6897247 │ []          │                  0 │         0 │          0 │                 0 │     0 │           │
└────────────┴────────────┴─────────────────────┴─────────────┴────────────────────┴──────────────┴───────────────────────────────────────────────────────────────────────────────────────┴──────┴───────────────┴─────────────┴────────────────────┴───────────┴────────────┴───────────────────┴───────┴───────────┘

1 rows in set. Elapsed: 0.335 sec. Processed 31.11 million rows, 2.00 GB (92.84 million rows/s., 5.97 GB/s.)
```
```
SELECT * EXCEPT database EXCEPT table EXCEPT partition
FROM system.parts
WHERE name = '15-4-0_3021_3021_0'

Query id: 0e0ebe32-01c9-4e2f-a2ce-c6fef2b46e45

┌─name───────────────┬─────────────────────────────────uuid─┬─part_type─┬─active─┬─marks─┬─rows─┬─bytes_on_disk─┬─data_compressed_bytes─┬─data_uncompressed_bytes─┬─marks_bytes─┬───modification_time─┬─────────remove_time─┬─refcount─┬───min_date─┬───max_date─┬────────────min_time─┬────────────max_time─┬─partition_id─┬─min_block_number─┬─max_block_number─┬─level─┬─data_version─┬─primary_key_bytes_in_memory─┬─primary_key_bytes_in_memory_allocated─┬─is_frozen─┬─engine───────────────────────┬─disk_name─┬─path──────────────────────────────────────────────────────────────────────────────────┬─hash_of_all_files────────────────┬─hash_of_uncompressed_files───────┬─uncompressed_hash_of_compressed_files─┬─delete_ttl_info_min─┬─delete_ttl_info_max─┬─move_ttl_info.expression─┬─move_ttl_info.min─┬─move_ttl_info.max─┬─default_compression_codec─┬─recompression_ttl_info.expression─┬─recompression_ttl_info.min─┬─recompression_ttl_info.max─┐
│ 15-4-0_3021_3021_0 │ 00000000-0000-0000-0000-000000000000 │ Compact   │      1 │     4 │   77 │       6897247 │               6896099 │                36829018 │         928 │ 2021-02-27 00:11:54 │ 1970-01-01 00:00:00 │        5 │ 1970-01-01 │ 1970-01-01 │ 1970-01-01 00:00:00 │ 1970-01-01 00:00:00 │ 15-4-0       │             3021 │             3021 │     0 │         3021 │                         249 │                                   584 │         0 │ ReplicatedReplacingMergeTree │ disk12    │ /disk12/clickhouse/store/832/83200d96-3224-40dc-83de-ab7adedbf1ec/15-4-0_3021_3021_0/ │ 242454d627d311e6258a92383c0110c9 │ 5399a1fb04391ab6a486e28d5a9d95dc │ 27a76ceec4e170d6bb63b4ba6754d39d      │ 1970-01-01 00:00:00 │ 1970-01-01 00:00:00 │ []                       │ []                │ []                │ LZ4                       │ []                                │ []                         │ []                         │
└────────────────────┴──────────────────────────────────────┴───────────┴────────┴───────┴──────┴───────────────┴───────────────────────┴─────────────────────────┴─────────────┴─────────────────────┴─────────────────────┴──────────┴────────────┴────────────┴─────────────────────┴─────────────────────┴──────────────┴──────────────────┴──────────────────┴───────┴──────────────┴─────────────────────────────┴───────────────────────────────────────┴───────────┴──────────────────────────────┴───────────┴───────────────────────────────────────────────────────────────────────────────────────┴──────────────────────────────────┴──────────────────────────────────┴───────────────────────────────────────┴─────────────────────┴─────────────────────┴──────────────────────────┴───────────────────┴───────────────────┴───────────────────────────┴───────────────────────────────────┴────────────────────────────┴────────────────────────────┘

1 rows in set. Elapsed: 20.739 sec. Processed 556.34 thousand rows, 341.05 MB (26.83 thousand rows/s., 16.44 MB/s.)
```

On some other hosts, the parts mentioned in the replication queue error message existed before, but were merged into other parts later.

It seems as we insert more and more rows into the table, there are more and more "lost" parts, and it is causing the replication queue grow more and more. Stopping the writes improves the situation, but very slowly.

* Which ClickHouse server version to use

21.2.3.15

**Comment by filimonov** [comment]
Can you explain where those REPLACE_RANGE come from? Do you use smth like replace partition commands? How often?

**Comment by cyberhuman** [comment]
We have `ALTER TABLE database.<temp_table_name> MOVE PARTITION <partitionN> TO TABLE database.table`. There are 100 partitions and each was flushed once a minute on average, hence +/-100 move partition requests a minute.

**Comment by amosbird** [comment]
PS. you can do `SELECT * EXCEPT (query_id, database, table) ...`

**Comment by filimonov** [comment]
> We have ALTER TABLE database.<temp_table_name> MOVE PARTITION <partitionN> TO TABLE database.table. There are 100 partitions and each was flushed once a minute on average, hence +/-100 move partition requests a minute.

So you rewrite 100 partition fully every minute.... 

Can you please:
1) do update to latest clickhouse version v21.2.5.5-stable (the issue may be related somehow to https://github.com/ClickHouse/ClickHouse/issues/20203 )
2) the queue is not executed because 'postpone_reason' = Not executing log entry queue-0000107981 of type REPLACE_RANGE for part  because another DROP_RANGE or REPLACE_RANGE entry are currently executing. 

you need to check on the queue elements which are currently executing, and try to figure out if smth wrong with THEM. 
```
SELECT * FROM system.replication_queue WHERE is_currently_executing
```

**Comment by cyberhuman** [comment]
Thank you for looking into it!

> So you rewrite 100 partitions fully every minute (sounds bad, maybe at least remove the partitioning, I guess the table is quite small).
We're not rewriting, we're appending 🤔 
The partitioning must be there because of the way the data will be queried later.

We'll try to upgrade as soon as possible, but meanwhile, the currently executing element is different every time:
```
SELECT * EXCEPT (database, table)
FROM system.replication_queue
WHERE (type = 'REPLACE_RANGE') AND (is_currently_executing = 1)
ORDER BY create_time ASC
LIMIT 5

Query id: be412696-ab29-458e-a457-1808fa2f0a63

┌─replica_name─┬─position─┬─node_name────────┬─type──────────┬─────────create_time─┬─required_quorum─┬─source_replica─┬─new_part_name─┬─parts_to_merge─┬─is_detach─┬─is_currently_executing─┬─num_tries─┬─last_exception─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───last_attempt_time─┬─num_postponed─┬─postpone_reason─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬──last_postpone_time─┬─merge_type─┐
│ 1            │   146834 │ queue-0000322873 │ REPLACE_RANGE │ 2021-02-28 16:15:03 │               0 │ 0              │               │ []             │         0 │                      1 │         3 │ Code: 234, e.displayText() = DB::Exception: Not found part 77-4-0_10713_10713_0 (or part covering it) neither source table neither remote replicas (version 21.2.3.15) │ 2021-03-03 00:45:12 │       1425225 │ Not executing log entry queue-0000322873 of type REPLACE_RANGE for part  because another DROP_RANGE or REPLACE_RANGE entry are currently executing. │ 2021-03-03 00:45:12 │            │
└──────────────┴──────────┴──────────────────┴───────────────┴─────────────────────┴─────────────────┴────────────────┴───────────────┴────────────────┴───────────┴────────────────────────┴───────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────┴───────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────┴────────────┘

1 rows in set. Elapsed: 3.197 sec. Processed 146.86 thousand rows, 74.25 MB (45.94 thousand rows/s., 23.22 MB/s.)
```
```
SELECT * EXCEPT (database, table)
FROM system.replication_queue
WHERE (type = 'REPLACE_RANGE') AND (is_currently_executing = 1)
ORDER BY create_time ASC
LIMIT 5

Query id: 9ddf0cb2-ad8d-4a2a-ab0f-fdc56770e7b0

┌─replica_name─┬─position─┬─node_name────────┬─type──────────┬─────────create_time─┬─required_quorum─┬─source_replica─┬─new_part_name─┬─parts_to_merge─┬─is_detach─┬─is_currently_executing─┬─num_tries─┬─last_exception─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───last_attempt_time─┬─num_postponed─┬─postpone_reason─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬──last_postpone_time─┬─merge_type─┐
│ 1            │   146833 │ queue-0000323156 │ REPLACE_RANGE │ 2021-02-28 16:18:03 │               0 │ 0              │               │ []             │         0 │                      1 │         3 │ Code: 234, e.displayText() = DB::Exception: Not found part 19-4-0_10750_10750_0 (or part covering it) neither source table neither remote replicas (version 21.2.3.15) │ 2021-03-03 00:48:10 │       1425529 │ Not executing log entry queue-0000323156 of type REPLACE_RANGE for part  because another DROP_RANGE or REPLACE_RANGE entry are currently executing. │ 2021-03-03 00:48:09 │            │
└──────────────┴──────────┴──────────────────┴───────────────┴─────────────────────┴─────────────────┴────────────────┴───────────────┴────────────────┴───────────┴────────────────────────┴───────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────┴───────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────┴────────────┘

1 rows in set. Elapsed: 3.651 sec. Processed 146.86 thousand rows, 74.25 MB (40.22 thousand rows/s., 20.34 MB/s.)
```
```
SELECT * EXCEPT (database, table)
FROM system.replication_queue
WHERE (type = 'REPLACE_RANGE') AND (is_currently_executing = 1)
ORDER BY create_time ASC
LIMIT 5

Query id: 0e203289-bafc-4ed4-8484-eb52a4cb069e

Ok.

0 rows in set. Elapsed: 2.433 sec. Processed 146.86 thousand rows, 74.25 MB (60.37 thousand rows/s., 30.52 MB/s.)
```
```
SELECT * EXCEPT (database, table)
FROM system.replication_queue
WHERE (type = 'REPLACE_RANGE') AND (is_currently_executing = 1)
ORDER BY create_time ASC
LIMIT 5

Query id: ddaaa4e5-ed93-4c7d-8e7d-9aba93f8cbcf

┌─replica_name─┬─position─┬─node_name────────┬─type──────────┬─────────create_time─┬─required_quorum─┬─source_replica─┬─new_part_name─┬─parts_to_merge─┬─is_detach─┬─is_currently_executing─┬─num_tries─┬─last_exception─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬───last_attempt_time─┬─num_postponed─┬─postpone_reason─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬──last_postpone_time─┬─merge_type─┐
│ 1            │   146834 │ queue-0000326513 │ REPLACE_RANGE │ 2021-02-28 16:50:19 │               0 │ 0              │               │ []             │         0 │                      1 │         3 │ Code: 234, e.displayText() = DB::Exception: Not found part 46-4-0_10796_10796_0 (or part covering it) neither source table neither remote replicas (version 21.2.3.15) │ 2021-03-03 01:14:34 │       1428332 │ Not executing log entry queue-0000326513 of type REPLACE_RANGE for part  because another DROP_RANGE or REPLACE_RANGE entry are currently executing. │ 2021-03-03 01:14:33 │            │
└──────────────┴──────────┴──────────────────┴───────────────┴─────────────────────┴─────────────────┴────────────────┴───────────────┴────────────────┴───────────┴────────────────────────┴───────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────┴───────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴─────────────────────┴────────────┘

1 rows in set. Elapsed: 3.492 sec. Processed 146.86 thousand rows, 74.25 MB (42.06 thousand rows/s., 21.27 MB/s.)
```

**Comment by filimonov** [comment]
Please post an update after version upgrade

**Comment by cyberhuman** [comment]
We did an upgrade today but AFAIK it didn't help. I will check in a while.
Is there anything else I could look at?

**Comment by cyberhuman** [comment]
I took a look and there is no improvement after the upgrade.
What I noticed by the way is that on the hosts in question and other hosts with the same issue there are too many parts (tens or hundreds of thousands). The number slowly goes down as (or if) they're merged. But the number of ongoing merges is very low (1 or 2, sometimes none). On the "good" hosts that don't have the issue at the moment, there are only a few thousands of parts and there are hundreds of merges ongoing in parallel. I'm not sure what is the cause and what is the consequence. Do the stalled REPLACE_RANGE entries in the replication queue block merges as well? Or some issue with merges cause it? 🤔

**Comment by cyberhuman** [comment]
I have enabled debug logging and it looks like the replication queue entries are retried but always fail. And the reason for the failure is missing parts. What's worse, those parts have been removed from the replica that had them. Even the 15-4-0_3021_3021_0 mentioned in the beginning of the report was removed on the same day later. The log message was:
```
/var/log/clickhouse.log.2
1629521:2021.03.02 18:23:13.178876 [ 155184 ] {} <Information> database_20210224.table_local (83200d96-3224-40dc-83de-ab7adedbf1ec): Activated part 15-4-0_3021_3021_0
1806219:2021.03.02 18:23:47.623356 [ 155062 ] {} <Warning> database_20210224.table_local (83200d96-3224-40dc-83de-ab7adedbf1ec): Source replica does not have part 15-4-0_3021_3021_0. Removing it from ZooKeeper.
```

**Comment by cyberhuman** [comment]
We decided to start the table from scratch, but with an important difference.
Previously the temporary tables had only one replica to make it simpler. Now we create both replicas.
The process is ongoing and I noticed a suspicious thing on a few occasions: there were hundreds of REPLACE_RANGE elements waiting in the replication queue, with the same error of missing chunk. And that missing chunk was actually present on another replica. In the case I checked more deeply, the chunk was fetched after ~13 minutes and the replication queue cleared up.

So:
1. Is that long replication time ok?
2. Why doesn't it work when moving parts from a table with one replica into a table with two replicas?

**Comment by cyberhuman** [comment]
I think #21264 may be related. We've just upgraded to a version that contains the fix and it looks good so far 🤞🏻

**Comment by cyberhuman** [comment]
Yes I think that was it. I don't see this issue anymore.

**Comment by Zhile** [comment]
Which version do you use now?

> I think #21264 may be related. We've just upgraded to a version that contains the fix and it looks good so far 🤞🏻

**Comment by cyberhuman** [comment]
21.3.2.5
